Implement workload redirects

mhutch opened the pull request:
Implements workload redirects as defined in https://github.com/dotnet/designs/pull/221

mhutch committed:
Eliminate redundant existence checks in workload resolver

mhutch committed:
Implement workload redirects

mhutch commented:
I haven't added tests yet, but this demonstrates the transparent redirect approach

mhutch committed:
Fix 'avaliable' typo

mhutch committed:
Allow deferred opening of workload manifest stream

mhutch committed:
Add workload resolver method to determine updated workloads

mhutch committed:
Replace workload TempDirResolver with more generic OverlayResolver

mhutch committed:
Improve diagnosability of manifest conflicts

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -369,21 +418,84 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadId workloadId)
+        /// <param name="advertisingManifestResolver">A resolver that composes the advertising manifests with the installed manifests that do not have corresponding advertising manifests</param>
+        /// <param name="existingWorkloads">The IDs of all of the installed workloads</param>
+        /// <returns></returns>
+        public IEnumerable<WorkloadId> GetUpdatedWorkloads(WorkloadResolver advertisingManifestResolver, IEnumerable<WorkloadId> installedWorkloads)
```
That is a new logic? WorkloadManifestUpdater.CalculateManifestUpdates is doing similar

mhutch replied:
This is to replace it. This one handles redirects and cross-manifest dependencies.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadDefinition.cs:
```diff
@@ -34,4 +43,14 @@ public enum WorkloadDefinitionKind
         Build
     }
+
+    public class WorkloadRedirect : BaseWorkloadDefinition
```
I need more time to thing about it. So far to me, it is transparent in the workload manifest sense. However, workload ID is used many places outside manifest reader. Mostly installation record, and UI for display. As an ID, it should be unique. And now the assumption breaks. 

Every time we compare workload id, we need to consider if they are just an alias. And when we iterate workloads, which "id" we should display?

The current code should not be able to resolve installation record mismatch. We need some kind of migration of installation record. But what if there is a downgrade?

We need to re-evaluate every place we have a workloadid.toString and comparison. In general, I don't like 2 ID can point to the same thing (nuget package id being inconsistently case sensitive/ in sensitive caused a lot bugs already). It is a lot for the caller to watch out. Maybe we could override "==" and "toString". But I still prefer to just have another layer of "display id" and never change the "true id"

mhutch committed:
Improve workload composition errors

Manifests have an "informational path" describing where they were
loaded from. The workload resolver includes the manifest IDs and
associated informational path in composition errors, so it's easier
to see what caused the conflict.

mhutch committed:
Nullability checks in workload manifest reader tests

mhutch committed:
Stronger typing of workload/pack ids in public API

IIRC WorkloadId used to be private but now it's public there's no
reason not to use it

wli3 commented:
@mhutch hows the progress?

mhutch committed:
Disable workload redirects for now

We are not fully certain that they are transparent and that they won't
cause unintended knock-on effects.

mhutch committed:
Fix available workloads returned by resolver

Platform checks should take restrictions on base workloads into account,
as clarified in https://github.com/dotnet/designs/pull/230.
This means that the CLI can no longer trivially display the RIDs for
the platforms on which the workloads are available. This could be added
back in future using the RID graph to coalesce values, but would probably
need additional work to ensure the values were user friendly.

The resolver no longer returns workload definitions. The ensures clients
use the resolver's APIs that work on the composed manifest data. Having
all code that deals with manifest data direct in a central location
helps keep behaviour consistent.

It also does not be return abstract or empty workloads as these
cannot be installed.

mhutch committed:
Include workload ID in exception when resolver cannot find it

mhutch committed:
Hide WorkloadResolver.GetPackPath in favor of ResolvePackPath

Ensures aliased pack IDs are always resolved

mhutch committed:
Improve WorkloadResolver.GetPacksInWorkload

* Better error for missing base workload
* Now returns information about where the pack was referenced
  so better errors can be provided
* Factored out a helper for enumerating base workloads

mhutch committed:
When suggesting workloads, only consider available workloads

sfoslund commented:
@mhutch is this PR ready for review?

mhutch commented:
@sfoslund all the existing commits are ready for review, yes. I have a couple more commits to come.

sfoslund reviewed (approved):
LGTM overall but I'm not sure why CI isn't running. Can we move this out of draft mode? I'd like to get this in this week if possible.

mhutch commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -97,15 +97,14 @@ private void LoadManifestsFromProvider(IWorkloadManifestProvider manifestProvide
+                    if(!string.Equals (manifestId, manifest.Id, StringComparison.OrdinalIgnoreCase))
                     {
-                        throw new Exception($"Duplicate workload manifest {manifestId}");
+                        throw new WorkloadManifestCompositionException($"Manifest '{manifestId}' from provider {manifestProvider} does not match payload id '{manifest.Id}'");
```
They probably should, yes. I didn't worry because it's possible they'll never be seen, and they aren't very useful to end-users. They're basically internal errors and indicate either the manifests themselves are broken or the combination of manifests is broken.

That said, a lot of the workload manifest errors are localized, so these probably should be too for consistency.

Perhaps the CLI should catch `WorkloadManifestCompositionException` and `WorkloadManifestFormatException` and report an error that says the installation (or nuget sources) is broken and suggests repairing or reinstalling.

marcpopMSFT commented:
@mhutch Is this ready to take out of draft and get CI running as I believe some work from others (Sarah and Will) is waiting on this PR?

mhutch committed:
Localize remaining workload composition errors

Also improves errors from unresolved redirects, and handles
redirects in a second pass so they're not dependent on order
of manifests and definitions

sfoslund commented:
@mhutch do you mind resolving the conflicts so we can get CI running and merge? Let me know if you need a hand, I'm happy to help.

mhutch commented:
@sfoslund sure, i can rebase & resolve. is this targeting the right branch?

sfoslund commented:
> is this targeting the right branch?

Yep!

mhutch committed:
Simplify WorkloadResolver.GetInstalledManifests

We now have a collection of the manifests that are loaded, so use
that rather than rescanning. This makes it behave correctly for
correct for overlay resolvers.

Also return an "Info" object to be consistent with other API. We
may want to add more information onto this object in future.

marcpopMSFT commented:
@mhutch , we'll be snapping for preview 7 tomorrow.  Do you think you can get the builds fixed and ready by EOD tomorrow so we can get this in?

mhutch committed:
Fix ResolvePackPath returning empty workload pack ids

Fallout from the switch from string to id structs

mhutch committed:
Check updated composition error messages in workload tests

mhutch committed:
Don't assume missing workload packs can always be satisfied

When getting a suggestion for workloads to satisfy missing packs,
first determine which cannot be installed on the current machine,
and report them with a specific error code NETSDK1178.

dsplaisted committed:
Fix error message argument indices

dsplaisted committed:
Fix workload mocks to use correct manifest IDs

This prevents it from erroring out because it loads the same manifest
twice in the workload update test.

dsplaisted commented:
@mhutch @sfoslund I fixed some issues with the tests here.  There are still more failing tests which I may be able to investigate later.

dsplaisted committed:
Update test now that mock workload manifest has right ID

mhutch committed:
Fix accidental removal that broke workload search tests

dsplaisted committed:
Generate correct error message when trying to install a workload which isn't supported for the current platform

dsplaisted commented:
@mhutch @sfoslund I think I've fixed the remainder of the test failures.

I replaced `IWorkloadResolver.IsWorkloadPlatformCompatible` with `IsPlatformIncompatibleWorkload` in order to support querying whether a workload which wasn't returned by `GetAvailableWorkloads` was a valid workload ID which wasn't available on the current platform, in order to generate the correct error message.

sfoslund reviewed (approved):
LGTM, a few nits from my previous review are still left but we can follow up on those later. @mhutch @dsplaisted can we merge this?

sfoslund reviewed (approved):
LGTM, a few nits from my previous review are still left but we can follow up on those later. @mhutch @dsplaisted can we merge this?

mhutch commented:
Sounds good to me :)

KalleOlaviNiemitalo commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -512,7 +510,7 @@ public PackInfo(string id, string version, WorkloadPackKind kind, string path, s
             /// The workload pack ID. The NuGet package ID <see cref="ResolvedPackageId"/> may differ from this.
             /// </summary>
-            public string Id { get; }
+            public WorkloadPackId Id { get; }
```
I believe this PackInfo.Id type change caused <https://github.com/dotnet/sdk/issues/20350>.
